orderbook/src/database/app_data.rs

```diff
@@ -0,0 +1,206 @@
+use super::*;
+use crate::conversions::*;
+use anyhow::Result;
+use futures::stream::TryStreamExt;
+use futures::StreamExt;
+use model::app_data::{AppData, MetaData};
+use primitive_types::{H160, H256};
+use std::borrow::Cow;
+
+#[async_trait::async_trait]
+pub trait AppDataStoring: Send + Sync {
+    async fn insert_app_data(&self, app_data: &AppData) -> Result<H256, InsertionError>;
+    fn app_data<'a>(&'a self, filter: &'a AppDataFilter) -> BoxStream<'a, Result<AppData>>;
+}
+
+/// Any default value means that this field is unfiltered.
+#[derive(Clone, Default)]
+pub struct AppDataFilter {
+    pub app_data_hash: Option<H256>,
+    pub referrer: Option<H160>,
+}
+
+#[derive(Debug)]
+pub enum InsertionError {
+    DuplicatedRecord(H256),
+    AnyhowError(anyhow::Error),
+    DbError(sqlx::Error),
+}
+
+impl From<sqlx::Error> for InsertionError {
+    fn from(err: sqlx::Error) -> Self {
+        Self::DbError(err)
+    }
+}
+
+impl From<anyhow::Error> for InsertionError {
+    fn from(err: anyhow::Error) -> Self {
+        Self::AnyhowError(err)
+    }
+}
+
+#[async_trait::async_trait]
+impl AppDataStoring for Postgres {
+    async fn insert_app_data(&self, app_data: &AppData) -> Result<H256, InsertionError> {
```

Comment:
Won't the original data be stored anywhere?
Since we are storing the hash and the contents, we won't be able to recover it unless it's stored raw.

Reply:
We are storing the app_data struct. And since I can reconstruct the original json from the app_data(modulo ordering of parameters), I think there is no need to store the original data

Reply:
But this is under the assumption that only the referral address is relevant.
If we store the version and kind (it's another discussion what we should store), you'd have to flatten that structure too.

IMHO, a key-value store would make more sense for this sort of data (think of Redis/ElasticSearch/etc) as the doc format is easily queryable as is.

I just looked up and it can be stored and retrieved on Postgresql https://stackoverflow.com/questions/39010601/extract-key-value-from-json-objects-in-postgres https://www.postgresql.org/docs/9.3/functions-json.html

Doing that way can even reduce the need to store the other fields.

What do you think of this approach?

Reply:
I don't like, as we can not query easily all appdata that refer to a specific referrer. This is needed to build good queries later.

If we need the version or the kind, then I will also store them as field in the database